### PR TITLE
[server] Use JWTs for Sign-in flow state WEB-365

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

* Stops using session storage to store the AuthFlow
* Instead, uses a signed JWT as part of the sign-in dance attached to the `state` param
* The `state` param is echoed back to us by the auth provider, and we extract our information from it

## Related Issue(s)

Part of WEB-365

## How to test

1. Preview
2. Sign-in with different providers, observe sign-in works

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - mp-server-70f37720f6</li>
	<li><b>🔗 URL</b> - <a href="https://mp-server-70f37720f6.preview.gitpod-dev.com/workspaces" target="_blank">mp-server-70f37720f6.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

**Comment by geropl** [comment]
Taking a look now...

File: components/server/src/auth/generic-auth-provider.ts [hunk 1]
```diff
@@ -258,7 +267,17 @@ export abstract class GenericAuthProvider implements AuthProvider {
         });
 
         const isAlreadyLoggedIn = request.isAuthenticated() && User.is(request.user);
-        const authFlow = AuthFlow.get(request.session);
+
+        const state = request.query.state;
+        if (!state) {
+            log.error(cxt, `(${strategyName}) No state present on callback request.`, { clientInfo });
```
Comment (geropl) [inline components/server/src/auth/generic-auth-provider.ts:273]: This case is already handled in line 294ff, incl. a metric. Why handling it here again? Plus, it seems to shadow the `isAlreadyLoggedIn` case.

Comment (easyCZ) [reply components/server/src/auth/generic-auth-provider.ts:273]: This checks if the query contained state param, whereas 294 checks the auth flow exists (auth flow is derived (verified) out of the state param.

I can add the metirc here as well

File: components/server/src/auth/login-completion-handler.ts [hunk 2]
```diff
@@ -48,9 +47,6 @@ export class LoginCompletionHandler {
                 });
             });
         } catch (err) {
-            // Clean up the session & avoid loops
-            await AuthFlow.clear(request.session);
```
Comment (geropl) [inline components/server/src/auth/login-completion-handler.ts:52]: I wonder if we want to do some cleanup in all the places we did `AuthFlow.clear` before. But I cannot think of any... we should always use the hostUrl to craft redirects (like here), so we should be fine as well. :+1:

Comment (easyCZ) [reply components/server/src/auth/login-completion-handler.ts:52]: Yeah, I had a look at refs to `.clear()` but couldn't find anything else to drop/adjust.

File: components/server/src/auth/generic-auth-provider.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/auth/generic-auth-provider.ts:273]: Ah, I now see where isAlreadyLoggedIn is coming from: From the DB, which i future won't make sense for. :+1:

Jep, having the metrics here would make sense I think.

**Comment by easyCZ** [comment]
> Registering https://gitlab.gitpod-dev.com gave me: `The requested scope is invalid, unknown, or malformed.`
> 
> The URL is: `https://gitlab.gitpod-dev.com/oauth/authorize?response_type=code&redirect_uri=https%3A%2F%2Fmp-server-70f37720f6.preview.gitpod-dev.com%2Fauth%2Fgitlab.gitpod-dev.com%2Fcallback&scope=api read_user&state=mp-server-70f37720f6%2CeyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.REDACTED&client_id=4f22c49052f41309658cec3ecab6397297c2cad4d133b2a59246cac715c9d752`
> 
> Note the space in `api read_user`, but I bet it's failing because of state. Not sure why, it should pass it as of the OAuth spec... 🤷 Maybe it does not like the format/length?

With the saas version of gitlab, I was able to sign in. Will try to investigate

**Comment by easyCZ** [comment]
> Registering [gitlab.gitpod-dev.com](https://gitlab.gitpod-dev.com) gave me: `The requested scope is invalid, unknown, or malformed.`

This turns out to be un-related to the changes in this PR. I've been able to reproduce this behaviour on another preview env without these changes.

It comes down to the App integration in Gitlab not having the `read_repositoriy` permission (but only the API permissions) which then returns this error. If you grant `read_repository`, the flow goes through correctly.

**Comment by AlexTugarev** [comment]
@easyCZ, could you check the logs of the past ~10 min, please? 

Requests like the following one are returned with 403 Forbidden. 

https://mp-server-70f37720f6.preview.gitpod-dev.com/api/authorize?returnTo=https%3A%2F%2Fmp-server-70f37720f6.preview.gitpod-dev.com%2Fcomplete-auth%3Fmessage%3Dsuccess&host=github.com&override=true&scopes=user:email

I could reproduce a couple of times by signing in, then trying to connect to a second provider on https://mp-server-70f37720f6.preview.gitpod-dev.com/user/integrations

**Comment by easyCZ** [comment]
@AlexTugarev 

Logs show the following:
```
WARN: "User is trying to connect a provider identity twice.
ERROR: JsonWebTokenError: invalid signature
```

Which providers did you use? I've been able to connect both Github and Gitlab and another custom one as well.

**Comment by easyCZ** [comment]
Screenshot 2023-05-23 at 9 40 49

**Comment by AlexTugarev** [comment]
@easyCZ, just found the cause.

> {"component":"server","severity":"INFO","time":"2023-05-23T07:38:25.519Z","context":{"userId":"2aeb7c38-9d08-4853-bf3e-..."},"message":"JSON RPC Request getOwnAuthProviders failed with user error: 470/\"You've been blocked.\"","payload":{"method":"getOwnAuthProviders","args":[null,{"_isCancelled":false}],"code":470,"message":"You've been blocked."}}

I tested with several providers, at least one of them having a non gitpod.io email address. Blocked accounts are checked apparently. 

Continue testing after redeploying.

File: components/server/src/auth/generic-auth-provider.ts [hunk 3]
```diff
@@ -258,7 +267,17 @@ export abstract class GenericAuthProvider implements AuthProvider {
...
+            response.redirect(
+                this.getSorryUrl(`No state was present on the authentication callback. Please try again.`),
+            );
+            return;
+        }
+
+        const authFlow = await this.parseState(state as string);
```
Comment (AlexTugarev) [inline components/server/src/auth/generic-auth-provider.ts:280]: Potential parse/verification errors should be caught here, redirecting to the default error page would be ok. 

To reproduce a dangling request, first log out, then navigate to https://github.com/login/oauth/authorize?response_type=code&redirect_uri=http%3A%2F%2Fpreview.gitpod-dev.com%2Fauth%2Fgithub%2Fcallback&scope=user%3Aemail&state=mp-server-70f37720f6%2CeyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9-bar&client_id=6025fa48a14288c318ad

**Review by AlexTugarev — APPROVED** [review]
Looks great! Thanks for tackling this!

Just a single remark for error handling: https://github.com/gitpod-io/gitpod/pull/17628/files#r1201724667

**Pull request merged**